active/0014-rolling-cluster-upgrade.md

```diff
@@ -75,31 +95,48 @@ Non-trivial changes to the schema should be performed in stages:
 
 #### Table migration
 
+`bpapi` application takes care of the static checks.
 Writing migration in a way that avoids whole-table locks is a complicated process, and should be done on a case-by-case basis.
 Migration process could utilize mnesia transactions to traverse the tables entry-by-entry and move the records to the new table, deleting the old record.
 The read code should read both tables.
 
 ### Static checks
 
-Annotations can be used to specify the API functions.
+Annotations can be used to specify the API functions:
 
 ```erlang
--module(foo_bpapi).
+-module(foo_api).
 
 -intruduced_in({foo/3, {5,0,0}}).
 foo(A, B, C) ->
-    rpc:call(?MODULE, do_foo, [A, B, C]).
+    rpc:call(foo_api_impl, foo, [A, B, C]).
 
 -intruduced_in({bar/3, {5,0,0}}).
 -deprecated_in({bar/3, {5,1,0}}).
 foo(A, B, C) ->
-    rpc:call(?MODULE, do_bar, [A, B, C]).
+    rpc:call(foo_api_impl, bar, [A, B, C]).
+```
+
+```erlang
+-module(foo_api_impl).
+
+foo(A, B, C) ->
+    ...
+
+foo(A, B, C) ->
+    ...
 ```
 
 The following static checks should be written:
 
 1. (xref): Only the functions in `*_bpapi` modules, that have `-introduced_in()` annotation, are allowed to call `rpc:call` and `gen_rpc:call` function
-1. Functions that have `deprecated_in` annotation are not removed earlier than two releases after the deprecation.
+1. Functions that have `deprecated_in` annotation are not removed earlier than two releases after the deprecation
+1. Each API function does an RPC to an existing function (xref check)
+1. Compatibility of the typespecs of the API modules:
```

Comment:
Now I realize that this check is less useful than it looks, because it only verifies that an old version of the code can call the new version of the code, but not the other way around.